src/v/serde/parquet/tests/shredder_test.cc

```diff
@@ -0,0 +1,233 @@
+ * by the Apache License, Version 2.0
+ */
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "serde/parquet/shredder.h"
+
+#include <gtest/gtest.h>
+
+#include <variant>
+
+using namespace serde::parquet;
+
+namespace {
+template<typename... Args>
+schema_element
+group_node(ss::sstring name, field_repetition_type rep_type, Args... args) {
+    chunked_vector<schema_element> children;
+    (children.push_back(std::move(args)), ...);
+    return {
+      .type = std::monostate{},
+      .repetition_type = rep_type,
+      .name = std::move(name),
+      .children = std::move(children),
+    };
+}
+
+schema_element leaf_node(
+  ss::sstring name,
+  field_repetition_type rep_type,
+  physical_type ptype,
+  logical_type ltype = logical_type{}) {
+    return {
+      .type = ptype,
+      .repetition_type = rep_type,
+      .name = std::move(name),
+      .logical_type = ltype,
+    };
+}
+
+template<typename... Args>
+struct_value record(Args... field) {
+    chunked_vector<struct_field> fields;
+    (fields.push_back(struct_field{std::move(field)}), ...);
+    return fields;
+}
+
+template<typename... Args>
+repeated_value list(Args... field) {
+    chunked_vector<repeated_element> fields;
+    (fields.push_back(repeated_element{std::move(field)}), ...);
+    return fields;
+}
+
+struct value_collector {
+    explicit value_collector(const schema_element& root) {
+        int32_t col_index = 0;
+        root.for_each([this, &col_index](const schema_element& node) {
+            int32_t my_index = col_index++;
+            if (!node.is_leaf()) {
+                // Skip inner nodes.
+                return;
+            }
+            column_mapping[my_index] = columns.size();
+            columns.emplace_back();
+        });
+    }
+
+    void operator()(shredded_value val) const {
+        auto it = column_mapping.find(val.schema_element_index);
+        ASSERT_NE(it, column_mapping.end()) << "invalid column index";
+        columns[it->second].push_back(std::move(val));
+    }
+
+    std::unordered_map<int32_t, size_t> column_mapping;
+    mutable chunked_vector<chunked_vector<shredded_value>> columns;
+};
+
+MATCHER_P3(IsVal, val, r, d, "") {
+    return fmt::format("{}", arg.val) == val && r == arg.rep_level
+           && d == arg.def_level;
+}
+
+} // namespace
+
+// NOLINTNEXTLINE(*internal-linkage*)
+void PrintTo(
+  const chunked_vector<chunked_vector<shredded_value>>& columns,
+  std::ostream* os) {
+    *os << "[\n";
+    for (const auto& column : columns) {
+        *os << "  [\n";
+        for (const auto& val : column) {
+            *os << fmt::format(
+              "    {{val: {}, r: {}, d: {}}}\n",
+              val.val,
+              val.rep_level,
+              val.def_level);
+        }
+        *os << "  ]\n";
+    }
+    *os << "]\n";
+}
+
+using testing::ElementsAre;
+
+// NOLINTBEGIN(*magic-number*)
+TEST(RecordShredding, ExampleFromDremelPaper) {
```

Comment:
I will work on finding some more examples and tests, but the paper example actually covers basically all of the cases, which is pretty cool.

Reply:
Okay I added the tests from the twitter blog on parquet, I think they are pretty good now :)

Reply:
🔥 🔥